build-logic/build.gradle.kts

```diff
@@ -6,14 +8,29 @@ repositories {
     mavenCentral()
 }
 
+kotlin {
+    jvmToolchain(20)
+}
+
+// `build-logic` can produce Java 20 compatible code, but only after the plugin is built with kotlin 1.9,
+// kotlin 1.9 has a fix for https://youtrack.jetbrains.com/issue/KT-57495/Add-JVM-target-bytecode-version-20
+// temporairly using 17, since this is the highest, preinstalled java version on the CI
+val targetJavaVersion = JavaVersion.VERSION_17
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(targetJavaVersion.majorVersion.toInt())
+}
+tasks.withType<KotlinCompile>().configureEach {
+    kotlinOptions.jvmTarget = targetJavaVersion.toString()
+}
```

Comment:
These few lines can be removed after you [bump Kotlin to 1.9](https://github.com/pinterest/ktlint/pull/2113). 
`build-logic` is an internal project and given we know the "actual" project is also built with <latest> java version it's fine to target <latest> java as well. 
As mentioned in the comment, I set it to Java 17, only because it is preinstalled on the CI

Reply:
These lines were missing from your first attempt: https://github.com/pinterest/ktlint/pull/1889, this is where the incompatibility came from